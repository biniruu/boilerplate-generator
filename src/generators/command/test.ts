import getCertainConditions from '@utils/certainConditions'
import type { SelectOptions } from '_types'

const getTestCommands = (configOptions: SelectOptions) => {
  const { hasCssModule, hasJest, hasNext, hasNuxt, hasReact, hasStorybook, hasTypescript, hasVue, hasWebpack } =
    getCertainConditions(configOptions)

  const testDevDependencies: string[] = []

  if (hasJest) {
    /**
     * @testing-library/dom (DOM Testing Library)
     * {@link https://testing-library.com/docs/dom-testing-library/intro/}
     *
     * @testing-library/jest-dom (jest-dom)
     * {@link https://github.com/testing-library/jest-dom#readme}
     *
     * @testing-library/user-event (user-event)
     * {@link https://github.com/testing-library/user-event#readme}
     *
     * jest (Jest)
     * {@link https://jestjs.io}
     *
     * jest-environment-jsdom
     * {@link https://github.com/jestjs/jest#readme}
     *
     * Jest Extended
     * {{@link https://jest-extended.jestcommunity.dev/docs}}
     *
     * jest-watch-typeahead
     * {@link https://github.com/jest-community/jest-watch-typeahead#jest-watch-typeahead}
     *
     * jsdom
     * {@link https://github.com/jsdom/jsdom#readme}
     *
     * msw (Mock Service Worker)
     * {@link https://mswjs.io}
     *
     * ts-jest
     * {@link https://kulshekhar.github.io/ts-jest/}
     * ts-jest is necessary when youâ€™re working in a development environment, since it provides types of jest configuration
     *
     *
     * If you have added 'jest', you don't need to add these packages: jest-resolve and jest-resolve-dependencies.
     * I recommend adding @types/jest even if it is not a TypeScript project because of its IntelliSense.
     */
    testDevDependencies.push(
      '@testing-library/dom',
      '@testing-library/jest-dom',
      '@testing-library/user-event',
      '@types/jest',
      'jest',
      'jest-environment-jsdom',
      'jest-extended',
      'jest-html-reporters',
      'jest-watch-typeahead',
      'jsdom',
      'msw',
      'ts-jest',
    )
    /**
     * identity-obj-proxy
     * {@link https://github.com/keyz/identity-obj-proxy#readme}
     */
    if (hasCssModule) {
      if (hasTypescript) {
        testDevDependencies.push('@types/identity-obj-proxy')
      }
      testDevDependencies.push('identity-obj-proxy')
    }
    /**
     * @testing-library/react (React Testing Library)
     * {@link https://testing-library.com/docs/react-testing-library/intro/})
     *
     * @testing-library/react-hooks (react-hooks-testing-library)
     * {@link https://github.com/testing-library/react-hooks-testing-library#readme}
     *
     * react-test-renderer (Test Renderer)
     * {@link https://legacy.reactjs.org/docs/test-renderer.html}
     */
    if (hasReact || hasNext) {
      if (hasStorybook) {
        testDevDependencies.push('@storybook/react', '@storybook/testing-library')
      }
      testDevDependencies.push(
        '@testing-library/react',
        '@testing-library/react-hooks',
        '@types/react-test-renderer',
        'react-test-renderer',
      )
    }
    /**
     * @testing-library/vue (Vue Testing Library)
     * {@link https://testing-library.com/docs/vue-testing-library/intro/}
     *
     * @vue/vue3-jest
     * {@link https://github.com/vuejs/vue-jest#readme}
     */
    if (hasNuxt || hasVue) {
      testDevDependencies.push('@testing-library/vue', '@vue/vue3-jest')
    }
    // TODO: Edit this jsdoc
    /**
     * @storybook/addon-a11y (Accessibility testing)
     * {@link XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX}
     *
     * @storybook/addon-actions (Action logger)
     * {@link XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX}
     *
     * @storybook/addon-essentials (Essential addons)
     * {@link XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX}
     *
     * @storybook/addon-interactions (Interact with components)
     * {@link XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX}
     *
     * @storybook/addon-links (Link to other stories)
     * {@link XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX}
     *
     * @storybook/builder-webpack5 (Webpack 5)
     * {@link XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX}
     *
     * @storybook/manager-webpack5 (Webpack 5 Storybook manager)
     * {@link XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX}
     */
    if (hasStorybook && !hasReact && !hasNext) {
      if (hasWebpack) {
        testDevDependencies.push('@storybook/builder-webpack5', '@storybook/manager-webpack5')
      }
      testDevDependencies.push(
        '@storybook/addon-a11y',
        '@storybook/addon-actions',
        '@storybook/addon-essentials',
        '@storybook/addon-interactions',
        '@storybook/addon-links',
      )
    }
  }

  return {
    testDevDependencies,
  }
}

export default getTestCommands
