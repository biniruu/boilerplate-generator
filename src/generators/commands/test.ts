import type { SelectOptions } from '_types'

const getTestCommands = (configOptions: SelectOptions) => {
  const hasJest = configOptions.jest
  const hasNext = configOptions.next
  const hasNuxt = configOptions.nuxt
  const hasPostcss = configOptions.postcss
  const hasReact = configOptions.react
  const hasTailwind = configOptions.tailwind
  const hasTypescript = configOptions.typescript
  const hasVue = configOptions.vue
  const hasCssModule = hasPostcss || hasTailwind

  const testDevDependencies: string[] = []

  if (hasJest) {
    if (hasCssModule) {
      if (hasTypescript) {
        testDevDependencies.push('@types/identity-obj-proxy')
      }
      testDevDependencies.push('identity-obj-proxy')
    }
    if (hasReact || hasNext) {
      testDevDependencies.push(
        '@testing-library/react',
        '@testing-library/react-hooks',
        '@types/react-test-renderer',
        'react-test-renderer',
      )
    }
    if (hasTypescript) {
      testDevDependencies.push('ts-jest')
    }
    if (hasNuxt || hasVue) {
      testDevDependencies.push('@testing-library/vue', '@vue/vue3-jest')
    }
    // If you have added 'jest', you don't need to add these packages: jest-resolve and jest-resolve-dependencies.
    // I recommend adding @types/jest even if it is not a TypeScript project because of its IntelliSense.
    testDevDependencies.push(
      '@testing-library/dom',
      '@testing-library/jest-dom',
      '@testing-library/user-event',
      '@types/jest',
      'jest',
      'jest-environment-jsdom',
      'jest-watch-typeahead',
      'jsdom',
      'msw',
    )
  }

  return {
    testDevDependencies,
  }
}

export default getTestCommands
