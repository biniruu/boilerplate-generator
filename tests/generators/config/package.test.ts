import generatePackageConfig from '@generators/config/package';

import { elemCode, options, setTextContent } from '../../setTextContent.test';

test('should return a configuration for package.json', () => {
  setTextContent({ content: generatePackageConfig(options) });

  expect(elemCode?.textContent).toMatchSnapshot();
});

test('should return a configuration for package.json with Gatsby.js', () => {
  options.gatsby = true;

  setTextContent({ content: generatePackageConfig(options) });

  expect(elemCode?.textContent).toMatchSnapshot();
});

test('should return a configuration for package.json with Express.js', () => {
  options.express = true;

  setTextContent({ content: generatePackageConfig(options) });

  expect(elemCode?.textContent).toMatchSnapshot();
});

test('should return a configuration for package.json with Jest', () => {
  options.jest = true;

  setTextContent({ content: generatePackageConfig(options) });

  expect(elemCode?.textContent).toMatchSnapshot();
});

test('should return a configuration for package.json with Next.js', () => {
  options.next = true;

  setTextContent({ content: generatePackageConfig(options) });

  expect(elemCode?.textContent).toMatchSnapshot();
});

test('should return a configuration for package.json with React.js', () => {
  options.react = true;

  setTextContent({ content: generatePackageConfig(options) });

  expect(elemCode?.textContent).toMatchSnapshot();
});

test('should return a configuration for package.json with Storybook', () => {
  options.storybook = true;

  setTextContent({ content: generatePackageConfig(options) });

  expect(elemCode?.textContent).toMatchSnapshot();
});

test('should return a configuration for package.json with TypeScript', () => {
  options.typescript = true;

  setTextContent({ content: generatePackageConfig(options) });

  expect(elemCode?.textContent).toMatchSnapshot();
});

test('should return a configuration for package.json with Webpack', () => {
  options.webpack = true;

  setTextContent({ content: generatePackageConfig(options) });

  expect(elemCode?.textContent).toMatchSnapshot();
});

test('should return a configuration for package.json with Next.js, TypeScript and Storybook', () => {
  options.next = true;
  options.typescript = true;
  options.storybook = true;

  setTextContent({ content: generatePackageConfig(options) });

  expect(elemCode?.textContent).toMatchSnapshot();
});
