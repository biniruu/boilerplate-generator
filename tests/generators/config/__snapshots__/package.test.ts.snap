// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return a configuration for package.json 1`] = `
"{
  "name": "my-site",
  "version": "0.1.0",
  "private": true,
  "description": "this-is-my-site",
  "license": "ISC",
  "author": "my-name",
  "main": "src/index.js",
  "keywords": [
    "node",
    "javascript"
  ],
  "scripts": {
    "dev:server": "nodemon ./server/index.ts",
    "lint": "eslint . && stylelint .",
    "update-packages": "yarn upgrade --latest && yarn sync-package-version",
    "sync-package-version": "npx syncyarnlock -ks"
  }
}"
`;

exports[`should return a configuration for package.json with Express.js 1`] = `
"{
  "name": "my-site",
  "version": "0.1.0",
  "private": true,
  "description": "this-is-my-site",
  "license": "ISC",
  "author": "my-name",
  "main": "src/index.js",
  "keywords": [
    "node",
    "javascript"
  ],
  "scripts": {
    "dev:server": "nodemon ./server/index.ts",
    "lint": "eslint . && stylelint .",
    "update-packages": "yarn upgrade --latest && yarn sync-package-version",
    "sync-package-version": "npx syncyarnlock -ks"
  }
}"
`;

exports[`should return a configuration for package.json with Gatsby.js 1`] = `
"{
  "name": "my-site",
  "version": "0.1.0",
  "private": true,
  "description": "this-is-my-site",
  "license": "ISC",
  "author": "my-name",
  "main": "src/index.js",
  "keywords": [
    "node",
    "javascript"
  ],
  "scripts": {
    "develop": "gatsby develop",
    "start": "gatsby develop",
    "build": "gatsby build",
    "serve": "gatsby serve",
    "clean": "gatsby clean",
    "typecheck": "tsc --noEmit",
    "dev:server": "nodemon ./server/index.ts",
    "lint": "eslint . && stylelint .",
    "update-packages": "yarn upgrade --latest && yarn sync-package-version",
    "sync-package-version": "npx syncyarnlock -ks"
  }
}"
`;

exports[`should return a configuration for package.json with Jest 1`] = `
"{
  "name": "my-site",
  "version": "0.1.0",
  "private": true,
  "description": "this-is-my-site",
  "license": "ISC",
  "author": "my-name",
  "main": "src/index.js",
  "keywords": [
    "node",
    "javascript"
  ],
  "scripts": {
    "dev:server": "nodemon ./server/index.ts",
    "test": "jest",
    "test:update": "jest -u",
    "test:watch": "jest --watch",
    "lint": "eslint . && stylelint .",
    "update-packages": "yarn upgrade --latest && yarn sync-package-version",
    "sync-package-version": "npx syncyarnlock -ks"
  }
}"
`;

exports[`should return a configuration for package.json with Next.js 1`] = `
"{
  "name": "my-site",
  "version": "0.1.0",
  "private": true,
  "description": "this-is-my-site",
  "license": "ISC",
  "author": "my-name",
  "main": "src/index.js",
  "keywords": [
    "node",
    "javascript"
  ],
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "eslint . && stylelint .",
    "update-packages": "yarn upgrade --latest && yarn sync-package-version",
    "sync-package-version": "npx syncyarnlock -ks"
  }
}"
`;

exports[`should return a configuration for package.json with Next.js, TypeScript and Storybook 1`] = `
"{
  "name": "my-site",
  "version": "0.1.0",
  "private": true,
  "description": "this-is-my-site",
  "license": "ISC",
  "author": "my-name",
  "main": "src/index.js",
  "keywords": [
    "node",
    "javascript"
  ],
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "eslint . && stylelint .",
    "storybook": "storybook dev -p 6006",
    "build-storybook": "storybook build",
    "update-packages": "yarn upgrade --latest && yarn sync-package-version",
    "sync-package-version": "npx syncyarnlock -ks"
  }
}"
`;

exports[`should return a configuration for package.json with React.js 1`] = `
"{
  "name": "my-site",
  "version": "0.1.0",
  "private": true,
  "description": "this-is-my-site",
  "license": "ISC",
  "author": "my-name",
  "main": "src/index.js",
  "keywords": [
    "node",
    "javascript"
  ],
  "scripts": {
    "dev": "vite",
    "build": "tsc && NODE_ENV=development vite build --mode development",
    "build:prod": "yarn lint && tsc && vite build",
    "preview": "vite preview",
    "lint": "eslint . && stylelint .",
    "update-packages": "yarn upgrade --latest && yarn sync-package-version",
    "sync-package-version": "npx syncyarnlock -ks"
  }
}"
`;

exports[`should return a configuration for package.json with Storybook 1`] = `
"{
  "name": "my-site",
  "version": "0.1.0",
  "private": true,
  "description": "this-is-my-site",
  "license": "ISC",
  "author": "my-name",
  "main": "src/index.js",
  "keywords": [
    "node",
    "javascript"
  ],
  "scripts": {
    "dev:server": "nodemon ./server/index.ts",
    "storybook": "storybook dev -p 6006",
    "build-storybook": "storybook build",
    "lint": "eslint . && stylelint .",
    "update-packages": "yarn upgrade --latest && yarn sync-package-version",
    "sync-package-version": "npx syncyarnlock -ks"
  }
}"
`;

exports[`should return a configuration for package.json with TypeScript 1`] = `
"{
  "name": "my-site",
  "version": "0.1.0",
  "private": true,
  "description": "this-is-my-site",
  "license": "ISC",
  "author": "my-name",
  "main": "src/index.js",
  "keywords": [
    "node",
    "javascript"
  ],
  "scripts": {
    "dev:server": "nodemon ./server/index.ts",
    "dev": "nodemon",
    "start": "yarn compile & nodemon",
    "compile": "rm -r dist ; npx tsc --project tsconfig.build.json",
    "lint": "eslint . && stylelint .",
    "update-packages": "yarn upgrade --latest && yarn sync-package-version",
    "sync-package-version": "npx syncyarnlock -ks"
  }
}"
`;

exports[`should return a configuration for package.json with Webpack 1`] = `
"{
  "name": "my-site",
  "version": "0.1.0",
  "private": true,
  "description": "this-is-my-site",
  "license": "ISC",
  "author": "my-name",
  "main": "src/index.js",
  "keywords": [
    "node",
    "javascript"
  ],
  "scripts": {
    "dev:server": "nodemon ./server/index.ts",
    "build": "webpack --mode=production --node-env=production",
    "build:dev": "webpack --mode=development",
    "build:prod": "webpack --mode=production --node-env=production",
    "watch": "webpack --watch",
    "serve": "webpack serve",
    "lint": "eslint . && stylelint .",
    "update-packages": "yarn upgrade --latest && yarn sync-package-version",
    "sync-package-version": "npx syncyarnlock -ks"
  }
}"
`;
